craft_parts/plugins/base.py

```diff
@@ -60,3 +62,40 @@ def out_of_source_build(self) -> bool:
     @abc.abstractmethod
     def get_build_commands(self) -> List[str]:
         """Return a list of commands to run during the build step."""
+
+
+class PluginModel(BaseModel):
+    """Model for plugins using pydantic validation.
+
+    Plugins with configuration properties can use pydantic validation to unmarshal
+    data from part specs. In this case, extract plugin-specific properties using
+    the :func:`extract_plugin_properties` helper.
+    """
+
+    class Config:
+        """Pydantic model configuration."""
+
+        validate_assignment = True
+        extra = "forbid"
+        allow_mutation = False
+        alias_generator = lambda s: s.replace("_", "-")  # noqa: E731
+
+
+def extract_plugin_properties(
+    data: Dict[str, Any], *, plugin_name: str
+) -> Dict[str, Any]:
+    """Obtain plugin-specifc entries from part properties.
+
+    :param data: A dictionary containing all part properties.
+    :plugin_name: The name of the plugin.
+
+    :return: A dictionary with plugin properties.
+    """
+    plugin_data: Dict[str, Any] = {}
+    prefix = f"{plugin_name}-"
```

Comment:
So apparently I lied when I said this was true for v2. The python plugin has 'requirements' and 'constraints' that don't follow this model, and perhaps there are others...

Reply:
Might be worth just dropping this generalization and make each plugin specify the list explicitly.

Reply:
the reason is legacy, it is nice to know what is consumed by the plugin and what is consumed by the rest of the machinery

Reply:
I verified current v2 snapcraft plugins, `requirements` and `constraints` in the python plugin are the only properties that violate the name prefix rule. Craft-parts could establish that as a hard rule (and legacy cases could be pre-processed if needed?)

Reply:
Seems fair.  It looks like it's under control of the plugins anyhow so it should be OK even as-is?